### ISSUE TITLE
Add option for multiple source case to apply GL shaders to every frame separately and produce custom compositing

### ISSUE THREAD

**Issue by Demin92** [androidx/media#840]
Hello! 
You have [demo](https://github.com/androidx/media/tree/release/demos/gl) that shows how to apply shaders to texture that we get from decoder by surface and surfaceTexture. But in case with several merged mediasources( if I understood correctly) we will get single texture after compositing. But it will be grate to have option to apply different shaders on frames from different sources and then create custom compositing.

**Comment by droid-girl** [comment]
I would like to clarify, are you trying to build a composition of videos, you want to apply different effects to each asset and preview this composition?

**Comment by Demin92** [comment]
yes, also I would like to get texture from each asset and blend these textures as I prefer in shaders

**Comment by droid-girl** [comment]
@Demin92 : thank you for clarifying. 
Have you looked into [`setVideoEffects`](https://developer.android.com/reference/androidx/media3/exoplayer/ExoPlayer#setVideoEffects(java.util.List%3Candroidx.media3.common.Effect%3E))? You can use this functionality and create custom effects by extending `GlEffect`. You can see already available effects in the effect module or create a custom effect.
`setVideoEffects` covers most of the use cases that GL ExoPlayer demo covered and might be a better fit for your use case?

**Comment by Demin92** [comment]
As far as I understood In case of several merged assets [setVideoEffects](https://developer.android.com/reference/androidx/media3/exoplayer/ExoPlayer#setVideoEffects(java.util.List%3Candroidx.media3.common.Effect%3E)) allow me to apply effect on resulting texture obtained after blending multiple textures from different sources. But I need option to apply different custom effects to each texture from several source assets, and also I need option to customize texture blending

**Comment by dway123** [comment]
The demo you linked to before is for outputting on ExoPlayer, using a custom GL shader and one input video plus one input image. ExoPlayer.setVideoEffects can allow you do apply custom or media3.effect `GlEffect` instances, but also on only one input video.

If your goal is to edit and export a video, you can have custom effects on each asset by implementing the `GlEffect` interface and implementing your own shaders. We don't currently support injecting a custom `VideoCompositor`, but that would be required for you to customize texture blending past what's already supported in VideoCompositorSettings.

If your goal is to preview a video using exoplayer, we're working on a way to display a composition (>1 input video), but this is not yet available. 

In the meantime, I don't think we have a simple solution to do this general issue, of custom compositors and previewing compositions. If you'd really like this, you could have exoPlayer.setVideoEffects take as input a TextureOverlay or your custom GlEffect (with 1+ inputs), and use that GlEffect to do your custom texture blending, between ExoPlayer's main input video, and a secondary input video rendered to the overlay.

**Comment by Demin92** [comment]
@dway123 Thanks for your answer. 
Can you say approximately when the compositing preview feature plans to be implemented?

**Comment by dway123** [comment]
While we are actively working on features that will lead to this, it is quite a large feature. I expect this will be implemented sometime after Q3 2024.

**Comment by droid-girl** [comment]
Since 1.4 release, you can start experimenting with CompositionPlayer API to preview `Composition`s. Check out [Composition demo](https://github.com/androidx/media/tree/release/demos/composition) for more info. These APIs are still in development

**Comment by Demin92** [comment]
@droid-girl Thanks for info. But I need option to preview composition with two video sequences with cross timeranges to make custom transitions, for example. Am I correct that now  CompositionPlayer not supported it ?

**Comment by droid-girl** [comment]
We are working on APIs that will enable cross-fades between items. Feel free to file another github issue for it specifically and i will update it when it is available

**Comment by Demin92** [comment]
Looks like this issue already exists https://github.com/androidx/media/issues/1488
